### ISSUE TITLE
Encoding woes with s///e

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6442]
Migrated from [rt.perl.org#21948](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=21948) (status was 'resolved')

Searchable as RT21948$

**Comment by p5pRT** [comment]
### From alankila@Elma.Net

#### Created by alankila@korppi.elma.fi



Is this intended behaviour\, or a bug?

$str = "test\_me äö";
$str2 = $str;
my $break = decode\("ISO\-8859\-15"\, "¤"\); \# Euro in unicode
$str =~ s/\./$break/;
$str2 =~ s/\./$break/e;
print $str;  \# prints 3 wide chars like 'â¬est\_me Ã¤Ã¶'
print $str2; \# only prints one wide character&#8203;: 'â¬est\_me äö'

Something stinks here\.

If I now try to encode the $str and $str2 back to ISO\-8859\-15\,
$str comes out ok with euro symbol\, but the $str2 has lost äö
because the string was promoted to utf8 representation but the
pre\-existing characters were not encoded appropriately\!

print encode\("ISO\-8859\-15"\, $str\);  \# prints ¤est\_me äö
print encode\("ISO\-8859\-15"\, $str2\); \# prints ¤est\_me ?

My workaround&#8203;: decode $str2\, or encode $break within the expression
before the substitution takes place\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.8.0:

Configured by alankila at Thu Aug 22 14:13:36 EEST 2002.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.18, archname=i686-linux-thread-multi-ld
    uname='linux korppi.elma.fi 2.4.18 #1 smp mon may 20 00:01:31 eest 2002 i686 unknown '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=define
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm',
    optimize='-O2',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -I/usr/local/include -I/usr/include/gdbm'
    ccversion='', gccversion='2.96 20000731 (Red Hat Linux 7.1 2.96-85)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='long double', nvsize=12, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldl -lm -lpthread -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lpthread -lc -lcrypt -lutil
    libc=/lib/libc-2.2.4.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.2.4'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.0:
    /home/alankila/private_html/devel/lib
    /opt/perl5/5.8.0-thread/lib/5.8.0/i686-linux-thread-multi-ld
    /opt/perl5/5.8.0-thread/lib/5.8.0
    /opt/perl5/5.8.0-thread/lib/site_perl/5.8.0/i686-linux-thread-multi-ld
    /opt/perl5/5.8.0-thread/lib/site_perl/5.8.0
    /opt/perl5/5.8.0-thread/lib/site_perl
    .


Environment for perl v5.8.0:
    HOME=/home/alankila
    LANG=C
    LANGUAGE (unset)
    LC_CTYPE=fi_FI@euro
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/bin:/usr/bin:/usr/X11R6/bin:/usr/local/bin
    PERL5LIB=/home/alankila/private_html/devel/lib
    PERL_BADLANG (unset)
    SHELL=/usr/bin/zsh

-- 
alankila@elma.net (Antti Lankila, P. +358 50 386 6217)
Platform Manager // Elma Oyj Electronic Trading


```
</details>

**Comment by p5pRT** [comment]
### From @andk

>>>>> On 12 Apr 2003 00&#8203;:15&#8203;:24 \-0000\, Antti Lankila \(via RT\) \<perlbug\-followup@&#8203;perl\.org> said&#8203;:

&nbsp;  > \# New Ticket Created by  Antti Lankila 
&nbsp;  > \# Please include the string&#8203;:  \[perl \#21948\]
&nbsp;  > \# in the subject line of all future correspondence about this issue\. 
&nbsp;  > \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt2/Ticket/Display\.html?id=21948 >


&nbsp;  > This is a bug report for perl from alankila@&#8203;korppi\.elma\.fi\,
&nbsp;  > generated with the help of perlbug 1\.34 running under perl v5\.8\.0\.


&nbsp;  > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
&nbsp;  > \[Please enter your report here\]

&nbsp;  > Is this intended behaviour\, or a bug?

&nbsp;  > $str = "test\_me äö";
&nbsp;  > $str2 = $str;
&nbsp;  > my $break = decode\("ISO\-8859\-15"\, "¤"\); \# Euro in unicode
&nbsp;  > $str =~ s/\./$break/;
&nbsp;  > $str2 =~ s/\./$break/e;
&nbsp;  > print $str;  \# prints 3 wide chars like 'â¬est\_me Ã¤Ã¶'
&nbsp;  > print $str2; \# only prints one wide character&#8203;: 'â¬est\_me äö'

This bug has been fixed in the development track \(patch 18764\)\.

&nbsp;  > Something stinks here\.

&nbsp;  > If I now try to encode the $str and $str2 back to ISO\-8859\-15\,
&nbsp;  > $str comes out ok with euro symbol\, but the $str2 has lost äö
&nbsp;  > because the string was promoted to utf8 representation but the
&nbsp;  > pre\-existing characters were not encoded appropriately\!

&nbsp;  > print encode\("ISO\-8859\-15"\, $str\);  \# prints ¤est\_me äö
&nbsp;  > print encode\("ISO\-8859\-15"\, $str2\); \# prints ¤est\_me ?

It seems\, this was only a consequence of the above bug\. With bleadperl
both give the same output\.

&nbsp;  > My workaround&#8203;: decode $str2\, or encode $break within the expression
&nbsp;  > before the substitution takes place\.

\-\- 
andreas

**Comment by p5pRT** [comment]
### From dankogai@dan.co.jp

On Saturday\, April 12\, 2003\, at 09&#8203;:15  AM\, Antti Lankila \(via RT\) wrote&#8203;:
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
>
> Is this intended behaviour\, or a bug?
>
> $str = "test\_me äö";
> $str2 = $str;
> my $break = decode\("ISO\-8859\-15"\, "¤"\); \# Euro in unicode
> $str =~ s/\./$break/;
> $str2 =~ s/\./$break/e;
> print $str;  \# prints 3 wide chars like 'â¬est\_me Ã¤Ã¶'
> print $str2; \# only prints one wide character&#8203;: 'â¬est\_me äö'
>
> Something stinks here\.
>
> If I now try to encode the $str and $str2 back to ISO\-8859\-15\,
> $str comes out ok with euro symbol\, but the $str2 has lost äö
> because the string was promoted to utf8 representation but the
> pre\-existing characters were not encoded appropriately\!

Questions&#8203;:

\* IN WHAT ENCODING was the script in question written?  If it is NOT IN 
ISO\-8859\-15 it is likely to fail because of the literals therein\.

\* Did you also try 'use encoding' pragma ?

\* Would you send me the file as MIME attachement so the literals are 
preserved ?

Dan the Encode Maintainer

**Comment by p5pRT** [comment]
### From alankila@Elma.Net

On 14 Apr 2003\, Dan Kogai wrote&#8203;:

> > $str = "test\_me äö";
> > $str2 = $str;
> > my $break = decode\("ISO\-8859\-15"\, "¤"\); \# Euro in unicode
> > $str =~ s/\./$break/;
> > $str2 =~ s/\./$break/e;
> > print $str;  \# prints 3 wide chars like 'â¬est\_me Ã¤Ã¶'
> > print $str2; \# only prints one wide character&#8203;: 'â¬est\_me äö'
> >
> > Something stinks here\.
> >
> > If I now try to encode the $str and $str2 back to ISO\-8859\-15\,
> > $str comes out ok with euro symbol\, but the $str2 has lost äö
> > because the string was promoted to utf8 representation but the
> > pre\-existing characters were not encoded appropriately\!
> 
> Questions&#8203;:
> 
> \* IN WHAT ENCODING was the script in question written?  If it is NOT IN 
> ISO\-8859\-15 it is likely to fail because of the literals therein\.

Apologies\, I wasn't clear on that\. The script was encoded in ISO\-8859\-15\,
meaning that ä \(auml\) and ö \(ouml\) and ¤ \(euro\) were all single bytes\.

> \* Did you also try 'use encoding' pragma ?

Where? In the sample program\, or in my real program? For the sample program\, it
causes the literal "test\_me äö" to be immediately encoded in UTF8 internally
and therefore the bug ceases to manifest\. However\, if I manually encode it as
ISO\-8859\-15 string and then let the program continue normally\, the bug still
appears\, as you would expect\.

About using encoding\.pm in the context of the bigger program I am working with&#8203;:

I can only say that it has "disastrous" results in everything for me\. I am not
specific on this one \-\- I tried using that like half a year ago when 5\.8\.0 was
fresh and I kept on getting mis\-encoded/decoded utf8 / bogus utf8 stuff in my
output\. I also remember I got occasional errors like

panic&#8203;: sv\_setpvn called with negative strlen at /opt/perl5/5\.8\.0\-thread/lib/5\.8\.0/i686\-linux\-thread\-multi\-ld/Encode\.pm line 140\.

I got that again today when I played a while with Encode\. I didn't hunt it
down\, so I can't offer a nice oneliner for which caused that\, though :\-\(

To continue my rant\, when I was still trying out encoding\, I discovered that
data which I got from DBD&#8203;::DB2 driver had the utf8 flag switched to wrong
position for some of the values which I got from the DB\. I don't remember
whether it was also one of these s///e problems\, but I remember that the
literals only looked right after I had done Encode&#8203;::\_utf8\_off\(\) or
Encode&#8203;::\_utf8\_on\(\) call \(I don't rememebr\, which\, unfortunately\)\.

Also\, I remember an incident when I was doing something like s/\(\.\.\.\)/my $val =
lc\($1\); \<does something with $val now>/ge and found out that inexplainably\,
$val contained stuff like 20c0 or something which definitely was not part of
the original data but looked more like hexadecimal dump of something\. I believe
this was already reported by Autrijus Tang\, who found buggy handling of the
input scalar in some of the lc\, uc\, lcfirst and ucfirst functions\. \(The bug was
already fixed in some of them\.\)

In short\, my experience with encoding\.pm was "massive breakage of everything"\,
with guts and fluids oozing out of gaping wounds I didn't understand why/how
they appeared\. I have every reason to believe the same is still true today\,
because the perl version is same and software is essentially the same\. My
conclusion is that the encoding\.pm is a great concept and in short it should
just work\, but currently the bugs in 5\.8\.0 make it unusable\.

> \* Would you send me the file as MIME attachement so the literals are 
> preserved ?

Yes\, certainly\. An implementation has been attached to this email\.

\-\- 
alankila@&#8203;elma\.net \(Antti Lankila\, P\. \+358 50 386 6217\)
Platform Manager // Elma Oyj Electronic Trading

**Comment by p5pRT** [comment]
### From alankila@Elma.Net

[perlbug.pl](https://rt-archive.perl.org/perl5/Ticket/Attachment/55169/41573/perlbug.pl)

**Comment by p5pRT** [comment]
### From @jhi

As noted\, this bug has been resolved\.  I'm marking the problem ticket as resolved\, too\.

**Comment by p5pRT** [comment]
@jhi - Status changed from 'new' to 'resolved'
